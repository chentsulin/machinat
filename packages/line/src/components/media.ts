import { makeUnitSegment } from '@machinat/core/renderer';
import { UnitSegment, FunctionOf } from '@machinat/core/renderer/types';
import { annotateLineComponent } from '../utils';
import { LineComponent, LineMessageSegmentValue } from '../types';

/**
 * @category Props
 */
export type VideoProps = {
  /** URL of video file (Max character limit: 1000) */
  originalContentUrl: string;
  /** URL of preview image (Max character limit: 1000) */
  previewImageUrl: string;
  /**
   * ID used to identify the video when Video viewing complete event occurs. If
   * you send a video message with trackingId added, the video viewing complete
   * event occurs when the user finishes watching the video.
   */
  trackingId?: string;
};

/** @internal */
const __Video: FunctionOf<LineComponent<
  VideoProps,
  UnitSegment<LineMessageSegmentValue>
>> = function Video(node, path) {
  const { originalContentUrl, previewImageUrl, trackingId } = node.props;

  return [
    makeUnitSegment(node, path, {
      type: 'video' as const,
      originalContentUrl,
      previewImageUrl,
      trackingId,
    }),
  ];
};
/**
 * Video sends an video message.
 * @category Component
 * @props {@link VideoProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#video-message).
 */
export const Video: LineComponent<
  VideoProps,
  UnitSegment<LineMessageSegmentValue>
> = annotateLineComponent(__Video);

/**
 * @category Props
 */
export type AudioProps = {
  /** URL of audio file (Max character limit: 1000) */
  originalContentUrl: string;
  /** Length of audio file (milliseconds) */
  duration: number;
};

/** @internal */
const __Audio: FunctionOf<LineComponent<
  AudioProps,
  UnitSegment<LineMessageSegmentValue>
>> = function Audio(node, path) {
  const { originalContentUrl, duration } = node.props;
  return [
    makeUnitSegment(node, path, {
      type: 'audio' as const,
      originalContentUrl,
      duration,
    }),
  ];
};
/**
 * Audio sends an audio message.
 * @category Component
 * @props {@link AudioProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#video-message).
 */
export const Audio: LineComponent<
  AudioProps,
  UnitSegment<LineMessageSegmentValue>
> = annotateLineComponent(__Audio);
