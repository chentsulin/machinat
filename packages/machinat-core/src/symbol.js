// @flow
export const MACHINAT_ELEMENT_TYPE = Symbol('machinat.element');

export const MACHINAT_FRAGMENT_TYPE = Symbol('machinat.element.fragment');

export const MACHINAT_PAUSE_TYPE = Symbol('machinat.element.pause');

export const MACHINAT_PROVIDER_TYPE = Symbol('machinat.element.provider');

export const MACHINAT_THUNK_TYPE = Symbol('machinat.element.thunk');

export const MACHINAT_RAW_TYPE = Symbol('machinat.element.raw');

export const MACHINAT_NATIVE_TYPE = Symbol('machinat.component.native');

export const MACHINAT_SERVICES_CONTAINER = Symbol(
  'machinat.services.container'
);

export const MACHINAT_SERVICES_INTERFACE = Symbol(
  'machinat.services.interface'
);

export const MACHINAT_SERVICES_PROVIDER = Symbol('machinat.services.provider');
