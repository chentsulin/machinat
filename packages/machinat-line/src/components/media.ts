import { unitSegment } from '@machinat/core/renderer';
import { annotateLineComponent } from '../utils';

export const Video = (node, path) => {
  const {
    url,
    originalContentUrl,
    previewURL,
    previewImageUrl,
    trackingId,
  } = node.props;

  return [
    unitSegment(node, path, {
      type: 'video',
      originalContentUrl: originalContentUrl || url,
      previewImageUrl: previewImageUrl || previewURL,
      trackingId,
    }),
  ];
};
annotateLineComponent(Video);

export const Audio = (node, path) => {
  const { url, originalContentUrl, duration } = node.props;
  return [
    unitSegment(node, path, {
      type: 'audio',
      originalContentUrl: originalContentUrl || url,
      duration,
    }),
  ];
};
annotateLineComponent(Audio);
