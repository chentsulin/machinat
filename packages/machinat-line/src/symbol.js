export const LINE_NAITVE_TYPE = Symbol.for('machinat.line.native');

export const API_PATH = Symbol.for('machinat.line.api_path');
export const HAS_BODY = Symbol.for('machinat.line.has_body');
export const THREAD_ID = Symbol.for('machinat.line.thread_id');

export const NO_BODY = Symbol.for('machinat.line.act.no_body');
