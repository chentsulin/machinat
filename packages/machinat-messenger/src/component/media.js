import { annotate, asNative, asUnit, hasEntry } from 'machinat-utility';

import {
  ATTACHED_FILE_DATA,
  ATTACHED_FILE_INFO,
  MESSENGER_NAITVE_TYPE,
} from '../symbol';
import { ENTRY_MESSAGES } from '../apiEntry';

const nativeMediaFactroy = (name, type) => {
  const container = {
    [name]: ({ url, reusable, attachmentId, data, fileInfo }) => [
      {
        message: {
          attachment: {
            type,
            payload: {
              url,
              is_reusable: reusable,
              attachment_id: attachmentId,
            },
          },
        },
        [ATTACHED_FILE_DATA]: data,
        [ATTACHED_FILE_INFO]: fileInfo,
      },
    ],
  };

  return annotate(
    asNative(MESSENGER_NAITVE_TYPE),
    hasEntry(ENTRY_MESSAGES),
    asUnit(true)
  )(container[name]);
};

export const Image = nativeMediaFactroy('Image', 'image');
export const Video = nativeMediaFactroy('Video', 'video');
export const Audio = nativeMediaFactroy('Audio', 'audio');
export const File = nativeMediaFactroy('File', 'file');
