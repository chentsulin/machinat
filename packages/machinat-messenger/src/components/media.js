import { unitSegment } from '@machinat/core/renderer';
import { annotateMessengerComponent } from '../utils';

const mediaFactory = (name, type) => {
  const container = {
    [name]: (node, path) => {
      const { url, reusable, attachmentId } = node.props;

      return [
        unitSegment(node, path, {
          message: {
            attachment: {
              type,
              payload: {
                url,
                is_reusable: reusable,
                attachment_id: attachmentId,
              },
            },
          },
        }),
      ];
    },
  };

  return annotateMessengerComponent(container[name]);
};

export const Image = mediaFactory('Image', 'image');
export const Video = mediaFactory('Video', 'video');
export const Audio = mediaFactory('Audio', 'audio');
export const File = mediaFactory('File', 'file');
