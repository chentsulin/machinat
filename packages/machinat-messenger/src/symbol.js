export const MESSENGER_NAITVE_TYPE = Symbol.for('machinat.messenger.native');
