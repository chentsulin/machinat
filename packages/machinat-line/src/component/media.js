import { annotate, asNative, asUnit } from 'machinat-utility';

import { LINE_NAITVE_TYPE } from '../symbol';

export const Video = ({
  url,
  originalContentUrl,
  previewURL,
  previewImageUrl,
}) => [
  {
    type: 'image',
    originalContentUrl: originalContentUrl || url,
    previewImageUrl: previewImageUrl || previewURL,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(true))(Video);

export const Audio = ({ url, originalContentUrl, duration }) => [
  {
    type: 'image',
    originalContentUrl: originalContentUrl || url,
    duration,
  },
];

annotate(asNative(LINE_NAITVE_TYPE), asUnit(true))(Audio);
