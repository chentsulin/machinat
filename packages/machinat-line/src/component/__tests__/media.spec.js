import Machinat from 'machinat';

import { Audio, Video } from '../media';

import { LINE_NAITVE_TYPE } from '../../symbol';

import render from './render';

it.each([Audio, Video].map(C => [C.name, C]))(
  '%s is valid native unit component',
  (_, Media) => {
    expect(typeof Media).toBe('function');

    expect(Media.$$native).toBe(LINE_NAITVE_TYPE);
    expect(Media.$$entry).toBe(undefined);
    expect(Media.$$unit).toBe(true);
  }
);

it.each(
  [
    <Audio url="https://..." duration={6666} />,
    <Video url="https://..." previewURL="https://..." />,
  ].map(e => [e.type.name, e])
)('%s render match snapshot', (_, mediaElement) => {
  expect(render(mediaElement).map(act => act.value)).toMatchSnapshot();
});
