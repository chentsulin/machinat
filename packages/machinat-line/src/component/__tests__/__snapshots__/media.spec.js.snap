// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audio render match snapshot 1`] = `
Array [
  Object {
    "duration": 6666,
    "originalContentUrl": "https://...",
    "type": "image",
  },
]
`;

exports[`Video render match snapshot 1`] = `
Array [
  Object {
    "originalContentUrl": "https://...",
    "previewImageUrl": "https://...",
    "type": "image",
  },
]
`;
